Xamarin.Forms.Core/NavigationPage.cs

```diff
@@ -154,18 +154,28 @@ public Task<Page> PopAsync()
 		public async Task<Page> PopAsync(bool animated)
 		{
 			var tcs = new TaskCompletionSource<bool>();
-			if (CurrentNavigationTask != null && !CurrentNavigationTask.IsCompleted)
+			try
 			{
-				var oldTask = CurrentNavigationTask;
-				CurrentNavigationTask = tcs.Task;
-				await oldTask;
+				if (CurrentNavigationTask != null && !CurrentNavigationTask.IsCompleted)
+				{
+					var oldTask = CurrentNavigationTask;
+					CurrentNavigationTask = tcs.Task;
+					await oldTask;
+				}
+				else
+					CurrentNavigationTask = tcs.Task;
+
+				var result = await PopAsyncInner(animated, false);
+				tcs.SetResult(true);
+				return result;
 			}
-			else
-				CurrentNavigationTask = tcs.Task;
+			catch (Exception)
+			{
+				CurrentNavigationTask = null;
+				tcs.SetCanceled();
```

Comment:
any reason you made this a Canceled? opposed to SetException and pass the Exception in?

Reply:
Consider the following scenario (as ridiculous as it might seem) of this methods all executed in the same fashion as the test reproduction case:
1. PopAsync
2. PopAsync
3. PushAsync

- First Pop resolves and removes page.
- Second Pop fails with _'Popped page does not appear on top of current navigation stack'_ and sets exception on `tcs`
- Push also fails with _'Popped page does not appear on top of current navigation stack'_

My reasoning was that queued operations that wait on the `tcs.Task` (now stored as `CurrentNavigationTask`) will never execute, and the exception they would get does not belong to that specific invocation. This includes the `PopToRootAsync` and `PushAsync` methods.